File: src/semantic_analysis/label_loops.fs
module Label_loops

open Ast.Untyped

let rec labelStatement current_label = function
  | Break _ ->
      (match current_label with
       | Some l -> Break l
       | None -> failwith "Break outside of loop")
  | Continue _ ->
      (match current_label with
       | Some l -> Continue l
       | None -> failwith "Continue outside of loop")
  | While(condition, body, id) ->
      let new_id = UniqueIds.makeLabel "while"
      Statement.While(condition = condition, body = labelStatement (Some new_id) body, id = new_id)
  | DoWhile(body, condition, id) ->
      let new_id = UniqueIds.makeLabel "do_while"
      DoWhile(body = labelStatement (Some new_id) body, condition = condition, id = new_id)
  | For(init, condition, post, body, id) ->
      let new_id = UniqueIds.makeLabel "for"
      Statement.For(init = init, condition = condition, post = post, body = labelStatement (Some new_id) body, id = new_id)
  | Compound blk -> Compound (labelBlock current_label blk)
  | If(condition, thenClause, elseClause) ->
      If(condition = condition, thenClause = labelStatement current_label thenClause, elseClause = Option.map (labelStatement current_label) elseClause)
  | (Null | Return _ | Expression _) as s -> s

and labelBlockItem current_label = function
  | Stmt s -> Stmt (labelStatement current_label s)
  | decl -> decl

and labelBlock current_label (Block b) =
  Block (List.map (labelBlockItem current_label) b)

let labelDecl = function
  | FunDecl fn ->
      FunDecl { fn with body = Option.map (labelBlock None) fn.body }
  | var_decl -> var_decl

let labelLoops (Program decls) = Program (List.map labelDecl decls)

File: src/semantic_analysis/LabelLoops.rs
//! Assign unique labels to loops and bind `break`/`continue` statements.
#![allow(non_snake_case)]

use super::{Ast::*, UniqueIds};

fn label_statement(counter: usize, current: Option<&str>, statement: Statement) -> (usize, Statement) {
    match statement {
        Statement::Break(_) => (counter, Statement::Break(current.expect("Break outside of loop").into())),
        Statement::Continue(_) => (counter, Statement::Continue(current.expect("Continue outside of loop").into())),
        Statement::While(condition, body, _) => {
            let (counter, id) = UniqueIds::make_label("while", counter);
            let (counter, body) = label_statement(counter, Some(&id), *body);
            (counter, Statement::While(condition, Box::new(body), id))
        }
        Statement::DoWhile(body, condition, _) => {
            let (counter, id) = UniqueIds::make_label("do_while", counter);
            let (counter, body) = label_statement(counter, Some(&id), *body);
            (counter, Statement::DoWhile(Box::new(body), condition, id))
        }
        Statement::For(init, condition, post, body, _) => {
            let (counter, id) = UniqueIds::make_label("for", counter);
            let (counter, body) = label_statement(counter, Some(&id), *body);
            (counter, Statement::For(init, condition, post, Box::new(body), id))
        }
        Statement::Compound(block) => { let (c, b) = label_block(counter, current, block); (c, Statement::Compound(b)) }
        Statement::If(condition, then_clause, else_clause) => {
            let (counter, then_clause) = label_statement(counter, current, *then_clause);
            let (counter, else_clause) = match else_clause {
                Some(e) => { let (c, e) = label_statement(counter, current, *e); (c, Some(Box::new(e))) }
                None => (counter, None),
            };
            (counter, Statement::If(condition, Box::new(then_clause), else_clause))
        }
        other => (counter, other),
    }
}

fn label_block(counter: usize, current: Option<&str>, block: Block) -> (usize, Block) {
    block.0.into_iter().fold((counter, Block(Vec::new())), |(counter, mut out), item| {
        let (counter, item) = match item {
            BlockItem::Stmt(s) => { let (c, s) = label_statement(counter, current, s); (c, BlockItem::Stmt(s)) }
            declaration => (counter, declaration),
        };
        out.0.push(item); (counter, out)
    })
}

/// Port of `labelLoops`: returns the updated unique-id counter and program.
pub fn labelLoops(counter: usize, program: UntypedProgram) -> (usize, UntypedProgram) {
    program.0.into_iter().fold((counter, UntypedProgram(Vec::new())), |(counter, mut out), declaration| {
        let (counter, declaration) = match declaration {
            Declaration::FunDecl(mut function) => if let Some(body) = function.body.take() {
                let (c, body) = label_block(counter, None, body); function.body = Some(body);
                (c, Declaration::FunDecl(function))
            } else { (counter, Declaration::FunDecl(function)) },
            declaration => (counter, declaration),
        };
        out.0.push(declaration); (counter, out)
    })
}

pub fn label_loops(counter: usize, program: UntypedProgram) -> (usize, UntypedProgram) { labelLoops(counter, program) }
